### PR TITLE
feat: Support YubiKeys in KeePassXC open mode

### LINKED ISSUE #3910: KeepassXC mode: "open" does not work with args ["--yubikey", "1:2343434"]
## Describe the bug

When KeepassXC mode is set to `open`, the `open` command is positioned after arguments (e.g., `/usr/local/bin/keepassxc-cli --no-password --yubikey 1:2343434 open <path>`) instead of the valid syntax (e.g., `/usr/local/bin/keepassxc-cli open --no-password --yubikey 1:2343434 <path>`). Due to this, the HMAC-SHA1 slot number (and key ID) is interpreted as command instead of being value of the argument.

## To reproduce

chezmoi.yml
```yml
keepassxc:
  database: "~/Desktop/KeePass.kdbx"
  args: ["--no-password", "--yubikey", "1:2343434"]
  mode: "open"
  prompt: false
```

## Expected behavior

Chezmoi should prompt me to unlock YubiKey once and remember the session (never prompt me again in this run).

## Output of command with the `--debug` flag

```console
$ chezmoi apply --debug
...
time=2024-08-20T20:38:01.042+02:00 level=INFO msg=Start cmd="/usr/local/bin/keepassxc-cli --no-password --yubikey 1:1097617 open /home/<user>/Desktop/KeePass.kdbx" start=2024-08-20T20:38:01.042+02:00 err=<nil>
chezmoi: etc/ssh/sshd_config: template: etc/ssh/sshd_config.tmpl:1:3: executing "etc/ssh/sshd_config.tmpl" at <keepassxcAttribute "sshd-config" "sshd_config">: error calling keepassxcAttribute: Invalid command 1:1097617.
```

## Output of `chezmoi doctor`

<details>

```console
$ chezmoi doctor

RESULT    CHECK                       MESSAGE
warning   version                     v2.52.1, built by nixpkgs
ok        latest-version              v2.52.1
ok        os-arch                     linux/amd64 (Arch Linux)
ok        uname                       Linux <hostname> 6.10.5-arch1-1 #1 SMP PREEMPT_DYNAMIC Thu, 15 Aug 2024 00:25:30 +0000 x86_64 GNU/Linux
ok        go-version                  go1.22.5 (gc)
ok        executable                  /nix/store/0nbd9r0j2kyzi6rlmpn65mydwm3whlk0-chezmoi-2.52.1/bin/chezmoi
ok        upgrade-method              sudo-upgrade-package
ok        config-file                 ~/.config/chezmoi/chezmoi.yaml, last modified 2024-08-20T20:36:55+02:00
error     source-dir                  open ~/.local/share/chezmoi: no such file or directory
ok        suspicious-entries          ~/.local/share/chezmoi: no such file or directory
error     working-tree                open ~/.local/share/chezmoi: no such file or directory
ok        dest-dir                    ~ is a directory
ok        umask                       022
ok        cd-command                  found /usr/bin/bash
ok        cd-args                     /usr/bin/bash
info      diff-command                not set
ok        edit-command                found ~/.local/share/devbox/global/default/.devbox/nix/profile/default/bin/codium
ok        edit-args                   ~/.local/share/devbox/global/default/.devbox/nix/profile/default/bin/codium
ok        git-command                 found ~/.local/share/devbox/global/default/.devbox/nix/profile/default/bin/git, version 2.45.2
ok        merge-command               found ~/.local/share/devbox/global/default/.devbox/nix/profile/default/bin/codium
ok        shell-command               found /usr/bin/bash
ok        shell-args                  /usr/bin/bash
info      age-command                 age not found in $PATH
ok        gpg-command                 found /usr/bin/gpg, version 2.4.5
info      pinentry-command            not set
info      1password-command           op not found in $PATH
info      bitwarden-command           bw not found in $PATH
info      bitwarden-secrets-command   bws not found in $PATH
info      dashlane-command            dcli not found in $PATH
info      doppler-command             doppler not found in $PATH
info      gopass-command              gopass not found in $PATH
ok        keepassxc-command           found ~/.local/share/devbox/global/default/.devbox/nix/profile/default/bin/keepassxc-cli, version 2.7.9
ok        keepassxc-db                ~/Desktop/digital-identity/KeePass.kdbx is a file
info      keeper-command              keeper not found in $PATH
info      lastpass-command            lpass not found in $PATH
info      pass-command                pass not found in $PATH
info      passhole-command            ph not found in $PATH
info      rbw-command                 rbw not found in $PATH
info      vault-command               vault not found in $PATH
info      vlt-command                 vlt not found in $PATH
info      secret-command              not set
```

</details>

## Additional info

Validated that the operation works with correct order of command and arguments:

```console
keepassxc-cli open --no-password --yubikey 1:2343434 <path>
```

### PR DESCRIPTION
**Pull request by twpayne**
Fixes #3910.

@mihakrumpestar would you be able to test this? tl;dr replace `keepassxc.args` with `keepassxc.openArgs` in your config file and it should work.

It's not easy for me to write a test for it because I don't use KeePassXC, don't have a YubiKey, and the "open" mode has two processes communicating with each other through a terminal.

### CODE REVIEW

**Review by halostatue — APPROVED** [review]
The changes look good to me, but I have the same problem as you for testing.

**Comment by mihakrumpestar** [comment]
Thanks for the quick response.

I tested it and the arguments and the command are in correct order, but unfortunately chezmoi does not wait for confirmation (user has to physically touch the device/key) to confirm authentication.

Here is the debug output:

```console
time=2024-08-21T00:13:30.841+02:00 level=INFO msg=Start cmd="/home/<user>/.local/share/devbox/global/default/.devbox/nix/profile/default/bin/keepassxc-cli open --no-password --yubikey 1:2343434 /home/<user>/Documents/digital-identity/KeePass.kdbx" start=2024-08-21T00:13:30.841+02:00 err=<nil>
chezmoi: .config/rclone/rclone.conf: template: dot_config/rclone/rclone.conf.tmpl:2:3: executing "dot_config/rclone/rclone.conf.tmpl" at <keepassxcAttribute "rclone-config" "rclone.conf">: error calling keepassxcAttribute: Please present or touch your YubiKey to continue.
```

Steps: Immediately after the message, chezmoi exits, while the keys confirmation light is on (meaning it is waiting for confirmation).
This does work when running the above command ("keepassxc-cli") standalone.

**Comment by twpayne** [comment]
> I tested it and the arguments and the command are in correct order, but unfortunately chezmoi does not wait for confirmation (user has to physically touch the device/key) to confirm authentication.

Any clue as to what options or output KeePassXC gives before it can accept further commands? The relevant code is at https://github.com/twpayne/chezmoi/blob/7f44957f0bbcdea196c2a6b0547f50d7aac6d7eb/internal/cmd/keepassxctemplatefuncs.go#L193-L268.

**Comment by mihakrumpestar** [comment]
> Any clue as to what options or output KeePassXC gives before it can accept further commands? The relevant code is at

When doing both `show` and `open` in standalone keepassxc-cli, both show `Please present or touch your YubiKey to continue.`.

So, how it gets to `Hardware key error: operation would block` is a mystery.

As for the code, I don't have any clues. I did look up, and it seems it is not that trivial, this thread (last comment in particular) might give some hints [https://github.com/keepassxreboot/keepassxc/issues/8969](https://github.com/keepassxreboot/keepassxc/issues/8969). Unfortunately, I am not that familiar with bash scripting to give an answer.

**Comment by mihakrumpestar** [comment]
I did test right now with the template command:

- Regular mode:

    ```yml
    keepassxc:
      database: "~/Documents/digital-identity/KeePass.kdbx"
      args: ["--no-password", "--yubikey", "1:1097617"]
      prompt: false
    ```
    
    ```console
    $ /usr/local/bin/chezmoi execute-template '{{ keepassxcAttribute "git-config" "gitconfig" }}' --verbose
    Hardware key error: operation would block
    ```
    
    Works as expected.

- Open mode:

    ```yml
    keepassxc:
      database: "~/Documents/digital-identity/KeePass.kdbx"
      openArgs: ["--no-password", "--yubikey", "1:1097617"]
      mode: "open"
      prompt: false
    ```
    
    ```
    $ /usr/local/bin/chezmoi execute-template '{{ keepassxcAttribute "git-config" "gitconfig" }}' --verbose
    chezmoi: template: arg1:1:3: executing "arg1" at <keepassxcAttribute "git-config" "gitconfig">: error calling keepassxcAttribute: Please present or touch your YubiKey to continue.
    ```
    
    Does not block execution.
    
    Now, based on this ` error calling keepassxcAttribute: Please present or touch your YubiKey to continue.` it treats this wait as error already.

**Comment by twpayne** [comment]
Thanks for the testing!

Note that chezmoi already uses `expect` to talk to `keepassxc-cli` in open mode. This means that chezmoi provides a virtual interactive terminal to `keepassxc-cli`.

The reason for this error message:

    error calling keepassxcAttribute: Please present or touch your YubiKey to continue.

is that chezmoi was expecting an `Enter password to unlock...` message from `keepassxc-cli`, but received the `Please present or touch...` message instead.

I've pushed an second commit to PR that makes chezmoi expect the `Please present or touch...` message, and also improved the error messages that chezmoi returns when it encounters something that it doesn't expect.

Would you be able to test this?

I don't have a recent YubiKey or use KeePassXC, so this might take a few iterations to get right with you doing the testing and me writing the code. I hope you're OK with that - I'd like to get this working :)

**Comment by mihakrumpestar** [comment]
I was trying to debug using latest commit:

```console
❯ /usr/local/bin/chezmoi --version
chezmoi version dev, commit 2b2a25d3b242b13b7245dc48d524489ca4d0130d, built at 2024-08-25T22:20:42Z

// I added some modifications on how the outputs are here
❯ /usr/local/bin/chezmoi execute-template '{{ keepassxcAttribute "git-config" "gitconfig" }}' --verbose
matched keepassxcPleasePresentOrTouchYourYubiKeyToContinueRx
not matched keepassxcPromptRx
chezmoi: template: arg1:1:3: executing "arg1" at <keepassxcAttribute "git-config" "gitconfig">: error calling keepassxcAttribute: "\r\n": unexpected prompt
```

The error `error calling keepassxcAttribute` happens on [this line](https://github.com/twpayne/chezmoi/blob/2b2a25d3b242b13b7245dc48d524489ca4d0130d/internal/cmd/keepassxctemplatefuncs.go#L274). I used `strconv.Quote(output)` to print escaped characters to it is properly readable in cmd.

**Comment by twpayne** [comment]
Thanks again for the testing!

I've updated this PR with new code which works locally for me.

My test config contains:

```toml
[keepassxc]
        database = "/home/twp/Passwords.kdbx"
        openArgs = ["--no-password", "--yubikey", "2:18281566"]
        mode = "open"
```

Example run:

```console
$ chezmoi execute-template '{{ keepassxc "chezmoi test" | toJson }}'
Please present or touch your YubiKey to continue.
{"Notes":"","Password":"'rW*b25hhsP91a.*bX#C","Tags":"","Title":"chezmoi test","URL":"","UserName":"","Uuid":"{2d7c6567-1f44-4139-bafb-5f886b2e08b7}"}%                       
```

Would you be able to test this version?

**Comment by mihakrumpestar** [comment]
Tested and can confirm that it works. And is much faster than non `open` mode when deploying all the dotfiles.

Would it be possible to have just `args` instead of having additional `openArgs`? It seems redundant and definitely confusing, since in the `keepassxc-cli`, `args` and `openArgs` are just `args`.

**Comment by mihakrumpestar** [comment]
Tested and works as before. Thank you.

**Comment by twpayne** [comment]
Great! I'll merge once the CI checks are complete. I'll likely do the next release of chezmoi towards the end of the week (the release cycle is roughly every two weeks).

**Pull request merged**
